### DISCUSSION TITLE
uiImageSliders not reacting on incoming Midi messages

### DISCUSSION THREAD

**Discussion by dobo365** [RomanKubiak/ctrlr#485] [Q&A]
Hi,
  I'm facing the issue that, in a panel, uiImageSliders not reacting on incoming Midi messages (uiButtons are directly working fine).
Using 5.3.201.
I found a work around which is deblocking the situation and it is to switch from Panel mode to Edit mode then back.
**Anyone having a clue of what the issue could be and why is that switch doing something?**

The main issue is that you cannot perform that operation when you have the executable version of the panel and thus you are stuck...
So, I must absolutely understand the reason and find a workaround.

I will try replacing my ctrlr.settings file. Maybe something weird in there...

**Comment by dobo365** [comment]
OK. I found the reason...

This is related to my UDO Super 6 editor.
The synth can work either in low (CC) or high (NRPN) resolution.
I added a button to switch from one to the other.

I configure the modulators Midi section with CC, CC number BUT also leave the 4 NRPN messages.
The reason is that I don't know how to add/remove those 4 lines in the _MidiMessageMultiList_ property.
Apparently, leaving them is making the CC message not working at panel startup (for me this is a bug; CC messages should just ignore that property).

I still want to try having only one version of the panel but otherwise the solution is just to make 2 versions of course.

So, the question is now: **do any of you know how to add/modify/remove the lines in the _MidiMessageMultiList_ property?**
The classical setPropertyString doesn't work but also doesn't give any error. Maybe another bug?
For the purpose of writing something, I tried:
`modDelayTime:getComponent():setPropertyString ("midiMessageMultiList", "CC,Direct,Direct,-1,-1,F0 00 F7")`

I wonder if this is maybe a valuetree property or ?

**Comment by dnaldoog** [comment]
Hi @dobo365!

You need to access **getMidiMessage** not **getComponent**:

-     panel:getModulatorByName("target"):getMidiMessage():setPropertyString("midiMessageType", "6") -- or 5 etc
-     panel:getModulatorByName("target"):getMidiMessage():setPropertyString("midiMessageSysExFormula", "F0 41 10 6A 12 03 00 14 56 MS LS F7")
-     panel:getModulatorByName("target"):getMidiMessage():setPropertyString("midiMessageMultiList", "F0 41 10 6A 12 03 00 14 56 3F xx F7:F0 41 10 6A 12 03 00 16 56 3F xx F7")
-     panel:getModulatorByName("target"):getMidiMessage():setPropertyString("midiMessageCtrlrNumber", "7")  -- volume

 ** Note: the colon creates a line break for the multi message list

You can switch between them using:
 - panel:getModulatorByName("target"):**getMidiMessage**():setPropertyString("midiMessageType", "6")

Where 6 = multiMessage, 5 = "sysex", 9= none and 0 = cc etc

Here's a panel with code taken from an abandoned panel for the JV1010 I was working on.

[Change MIDI message from lua_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/9595445/Change.MIDI.message.from.lua_1_0.zip)

**Reply by dobo365** [reply]
Thx for your answer John!
Of course!!! And I know and already use this in the same panel. Jesus...
I must have been tired when I did the test and just did a wrong copy/paste...

Now, in practice, switching to NRPN from CC and vice versa is not working fine.
You start in CC. The panel sends and reacts on received CC messages fine.
You switch to NRPN. The panel sends NRPN fine but don't react on received ones.

I will try with a panel sets as NRPN. Switch to CC but also enable my MidiMessageReceived method to listen to CC.
Maybe that combination will work as MidiMessageReceived will bypass the internal listening of received messages that is not working when you switch.

**Comment by dnaldoog** [comment]
Hi Dominique,

On reflection, I realised that of course you would know that very well, but it's such a tangled maize (Ctrlr) that getting lost or forgetting what one knows well is par for the course. ;0
[](url)
I might have written a panel that uses NRPN - I think the panel for the PLG-150-AN I did way back and it might contain lua to detect incoming NRPN - I'll have to check when I get home (or maybe I [abandoned that approach?](https://ctrlr.org/forums/topic/panel-to-send-simple-cc-nrpn-using-a-slide/#post-119137) ), but my first thought is that if NRPN is multi messaged maybe Ctrlr can't assign an incoming NRPN to a control like it can with normal CC messages?

**Reply by dobo365** [reply]
Actually it works fine when you change ALL the needed properties.

What I did wrong was to assume that the Multi message property was not taken into account when the Message type is CC (and it shouldn't; for me this is a bug).
So, I set it and kept it in the panel whatever CC or NRPN (yes, lazy boy).
This was causing the issue of CC messages not recognized by the panel even if the control was on CC.
I figured this out as I had other exactly identical uiImageSliders with just CC and not switching to NRPN, thus with the Multi message property being empty. They worked fine.

But OK. I learned again something (need to change all props to switch between CC and NRPN) and, before finding the issue/solution, I also built a MidiMessageReceived that captured CC messages. That worked fine even if, now, this will not be used.
I will look to your post as I'm wondering how to build an NRPN reader... 
Requires just a bit of coding I think: start a "thread" when seeing 99 and 98 that indicate the control; handle 6 and 38 as value. Same control until another 99,98 pair is received.
What I'm afraid is that Lua and the method cannot handle the amount of Midi messages received when the user moves a control from 20 to for example 15678 (if 0-16383).

Always nice exchanging with you!

**Reply by dnaldoog** [reply]
Yes always nice exchanging!

Well, the PLG-150-AN doesn't transmit MIDI so that panel has no midi received code, so it wasn't that. But I did find a small panel I must have been writing to Tedjuh about, because of the name, that has the following syntax if it helps, but you probably know all this anyway.

<pre>
CC,ByteValue,MSB7bitValue,99,-2:
CC,ByteValue,LSB7bitValue,98,-2:
CC,ByteValue,LSB7bitValue,6,-1:
CC,ByteValue,MSB7bitValue,101,16384:
CC,ByteValue,LSB7bitValue,100,127
</pre>

[tedjuh pan 1-nrpn.zip](https://github.com/RomanKubiak/ctrlr/files/9611834/tedjuh.pan.1-nrpn.zip)
